pkg/cmd/commandCollection.go

```diff
@@ -0,0 +1,351 @@
+	COMMAND_NAME_PROPERTIES_NAMESPACE     = "properties namespace"
+	COMMAND_NAME_PROPERTIES_NAMESPACE_GET = "properties namespace get"
+	COMMAND_NAME_RUNS                     = "runs"
+	COMMAND_NAME_RUNS_DOWNLOAD            = "runs download"
+	COMMAND_NAME_RUNS_GET                 = "runs get"
+	COMMAND_NAME_RUNS_PREPARE             = "runs prepare"
+	COMMAND_NAME_RUNS_SUBMIT              = "runs submit"
+	COMMAND_NAME_RUNS_SUBMIT_LOCAL        = "runs submit local"
+)
+
+// -----------------------------------------------------------------
+// Public functions.
+// -----------------------------------------------------------------
+func NewCommandCollection(factory Factory) (CommandCollection, error) {
+
+	commands := new(CommandCollectionImpl)
+
+	err := commands.init(factory)
+
+	return commands, err
+}
+
+// The main entry point into the cmd package.
+func Execute(factory Factory, args []string) error {
+	var err error
+
+	finalWordHandler := factory.GetFinalWordHandler()
+
+	var commands CommandCollection
+	commands, err = NewCommandCollection(factory)
+
+	if err == nil {
+
+		// Catch execution if a panic happens.
+		defer func() {
+			err := recover()
+
+			// Display the error and exit.
+			finalWordHandler.FinalWord(commands.GetRootCommand(), err)
+		}()
+
+		// Execute the command
+		err = commands.Execute(args)
+	}
+	finalWordHandler.FinalWord(commands.GetRootCommand(), err)
+	return err
+}
+
+func (commands *CommandCollectionImpl) GetRootCommand() GalasaCommand {
+	return commands.GetCommand(COMMAND_NAME_ROOT)
+}
+
+func (commands *CommandCollectionImpl) GetCommand(name string) GalasaCommand {
+	cmd, _ := commands.commandMap[name]
+	return cmd
+}
+
+func (commands *CommandCollectionImpl) Execute(args []string) error {
+
+	rootCmd := commands.GetRootCommand().GetCobraCommand()
+	rootCmd.SetArgs(args)
+
+	// Execute the command
+	err := rootCmd.Execute()
+
+	return err
+}
+
+// -----------------------------------------------------------------
+// Private functions.
+// -----------------------------------------------------------------
+func (commands *CommandCollectionImpl) init(factory Factory) error {
+
+	commands.commandMap = make(map[string]GalasaCommand)
+
+	rootCommand, err := NewRootCommand(factory)
+	if err == nil {
+		commands.rootCommand = rootCommand
+		commands.commandMap[rootCommand.GetName()] = rootCommand
+	}
+
+	if err == nil {
+		err = commands.addAuthCommands(factory, rootCommand)
+	}
+
+	if err == nil {
+		err = commands.addLocalCommands(factory, rootCommand)
+	}
+
+	if err == nil {
+		err = commands.addProjectCommands(factory, rootCommand)
+	}
+
+	if err == nil {
+		err = commands.addPropertiesCommands(factory, rootCommand)
+	}
+
+	if err == nil {
+		err = commands.addRunsCommands(factory, rootCommand)
+	}
+
+	if err == nil {
+		sanitiseCommandHelpDescriptions(rootCommand.GetCobraCommand())
+	}
+
+	return err
+}
+
+func (commands *CommandCollectionImpl) addAuthCommands(factory Factory, rootCommand GalasaCommand) error {
```

Comment:
Lots of repetition. We need an add function which takes a function as a parameter.